code/parachain/frame/assets-registry/src/lib.rs

```diff
@@ -220,20 +219,19 @@ pub mod pallet {
 		#[pallet::weight(<T as Config>::WeightInfo::register_asset())]
 		pub fn register_asset(
 			origin: OriginFor<T>,
-			local_or_foreign: LocalOrForeignAssetId<T::LocalAssetId, T::ForeignAssetId>,
+			protocol_id: [u8; 8],
+			nonce: u64,
+			location: Option<T::ForeignAssetId>,
 			asset_info: AssetInfo<T::Balance>,
 		) -> DispatchResult {
 			T::UpdateAssetRegistryOrigin::ensure_origin(origin)?;
 
-			let (asset_id, location) = match local_or_foreign {
-				LocalOrForeignAssetId::Local(asset_id) => (asset_id, None),
-				LocalOrForeignAssetId::Foreign(location) => (
-					T::LocalAssetId::from(u128::from_be_bytes(sp_io::hashing::blake2_128(
-						&location.encode(),
-					))),
-					Some(location),
-				),
-			};
+			let asset_id = Self::generate_asset_id(protocol_id, nonce);
```

Comment:
So foreign assets must have a protocol_id and a nonce now? strange

Reply:
Additional overhead for registering xcm assets?

Reply:
no hashing for xcmv2 as unstable -> some id -> why not prefix id -> no need for enum prefix (at least no need to store it).

Reply:
Ok any example? Say we want to create aUSD(assuming it's not already defined), how is the prefix and counter are calculated deterministically?

Reply:
For now, asset creation is a permissioned workflow. Meaning that we control the routes used for it and can enforce the protocol&nonce behavior with relative ease.

Supporting permissionless asset creation will need to come as its own feature. Once it's added, we can ensure routes used to permissionlessly create assets make the protocol&nonce behavior easy to use.

Having local asset IDs determined by two entirely different functions was technically erroneous as it did not technically ensure the lack of ID collisions without extra validation. Using protocol&nonce everywhere allows us to have predictable and  collision free asset IDs.

Reply:
>What do you mean deterministically?

This is an example provided by @KaiserKarel : A contract running on XCVM on a different chain calling into Picasso resulting in creation/output of an asset. Then requiring some amount of it to be transferred somewhere else. The second action requires to know the id of the asset. To know it that contract has to call Picasso again to know or read some output somehow. Hence the idea of determinism to know a generate assetId upfront using hashing based on XCM multiasset format. 

>Having local asset IDs determined by two entirely different functions was technically erroneous...

Could have been easily solved by just having hashing by using a known function as the last step of both schemes.

Reply:
> Supporting permissionless asset creation will need to come as its own feature. Once it's added, we can ensure routes used to permissionlessly create assets make the protocol&nonce behavior easy to use.

Concern I have is that if we are making it complicated to achieve that feature by forcing a certain scheme on generating assetIds already. The register asset interface that allows specifying `protocol_id` and `nonce` also allows intentional collisions to DOS assetID generation. I would be happy to resolve if this design addresses how to come up with protocol_ids and nonces for foreign assets while ensuring those can not be arbitrary chosen to intentionally create collisions.

Reply:
I believe you are correct. The original design was better and avoided collisions. Too much future proofing for unknown futures and unnecessary optimization has gone into this asset ID system as it is.

I'm going to re-introduce two definitive routes for local asset ID creation.

For local assets:

A protocol ID and nonce will be introduced for all locally created assets. If users ever need to create local assets, the protocol ID will be derived from the interface used to create the asset.

To avoid collisions, this will then be hashed. The hashing of this number is not a no-op, it is collision avoidance. The protocol and nonce only exists to generate a pre-image for hashing with input that can be known before the asset is created.

For foreign assets:

Foreign assets will be created with a hash of whatever location input was originally provided to create the asset. A different storage map will need to be introduced for each new location type (XCMv2, XCMv3, IBC Denom).

Exceptions:

To enable legacy assets that already have IDs and to allow for creating a new foreign asset that already exists in another location format, trait functions will expose the raw asset ID to be set as needed.

In the event that a foreign asset has already been created with another foreign location and needs to exists as the same local asset, the usage of multiple maps for different location types will enable us to use the `update_location` function to provide mappings between the various locations.

cc: @benluelo @dzmitry-lahoda @vimukthi-git

Reply:
i really dislike it and it does not add anything to determinism. i have described all flows in our chat. and all issues with xcm. i will not approve this design. so any failures are up to you than.

Reply:
like ai said, i was trying to be helpful by you are not listening. i am not aggressive, just assertive:)

Reply:
by failures i mean operational and governance nightmare and avoidable complication.

Reply:
Misunderstanding and not listening are two different things.
Can you express, in text, a situation in which the proposed system you're against would fail? @dzmitry-lahoda

Reply:
From @dzmitry-lahoda :
> 1. If asset id is one single global governed by local consensus then one may not achieve this ID predicatability if this ID is used as cross chain id
> 2. IBC-ICS20 and XCM do not care. They use counterp party prefixed/denom and inverted locations to have predicatbility. Will call these ids XCID
> 3. predictability = determism = when other consensus on other chain issues asset on this chain it know how to manage that access right away without queires assets id created (so IBC-ICS20 and XCM just do XCID != ID)
> 4. XCVM does not operates XCM or IBC-ICS20 XCIDs, so it should get its own predictability
> 5. It can easy to do so by having its satellite (native XCVM part on some chain running)  would manage its own ID and send it to this chain
> 6. Creating assets is easy, making bridge is high risk and cost governance operation.
> 7. So XCVM, when it creates bridge, will allow for that bridge to take protocol_prefix
> 8. So when other chain sends over XCVM asset creation, it sends its own under 8 bytes id. We just prefix it with protocol. Protocol is know to both parties as it was created during handshare (connection open, contract deploy, governance voting, whatever). So all things are spam protected. One consensus cannot spam assset id space of other consesus. In XCVM XCID = ID.
> 9. So on top of that hashes do not need for predictability, they complicate codebase, governance and migration with additional layers into already complicated topic.
> So what is use case for hashes?
> Why it is better?

cc: @vimukthi-git

Reply:
Proposal, use protocol/nonce for both local and foreign assets. In the event a contract or external consumer needs to create an asset, they should provide their own protocol ID and be responsible for generating a nonce.

To prevent a steeply unbalanced Merkle Tree, we can either

A) Hash the protocol/nonce to create the final local asset ID
  Pros:
  * Merkle tree should be balanced
  Cons:
  * More work to determine final asset IDs for consumers
B) Use concat keys
  Pros:
  * Merkle tree should be balanced
  Cons:
  * Migration would need more work as we'd be migrating the key format. Becomes exceptionally more difficult to migrate and ensure storage is correct.

Reply:
Perhaps in the future, external consumers could provide stake/purchase a protocol ID to create assets on our system. However, I don't think we can practically fit that feature into release 3.
Thoughts? @vimukthi-git @dzmitry-lahoda

Reply:
@dzmitry-lahoda 

> 2. .... They use counterp party prefixed/denom and inverted locations to have predicatbility.

What are prefixed/denom and inverted locations? Any example? Are you talking about [universal asset identifiers](https://github.com/paritytech/xcm-format#6-universal-asset-identifiers) spec?

> 3. .... predictability = determism = when other consensus on other chain issues asset on this chain it know how to manage that access right away without queires assets id created

This would just be one aspect I guess. There is also the aspect of preventing or authenticating that an asset received is actually the registered asset to prevent incorrect mapping or denial of service for registering assets. Say we received hypothetically 200 of IBC Osmosis chain asset of assetId=USDT. Then Picasso should understand how to correctly map it to already registered asset in the registry, say IBC/Osmosis/USDT. The channel which the asset arrives on should provide enough detail to map instead of being arbitrary. This is like signature verification of incoming messages on a channel to verify if a message is coming from the said source.

> ... XCVM does not operates XCM or IBC-ICS20 XCIDs, so it should get its own predictability

Agree. We want each protocol/pallet/chain to be able to generate their own independent predictable asset IDs. But we want to be able to convert them to a universally unique ID when registering on our chain to be able to map. 

> ... It can easy to do so by having its satellite (native XCVM part on some chain running) would manage its own ID and send it to this chain

Yes, though if the asset need to be transparent/accessible to substrate pallets they need to be stored in orml-tokens and registered in asset-registry.

> ... Creating assets is easy, making bridge is high risk and cost governance operation.
> ... So XCVM, when it creates bridge, will allow for that bridge to take protocol_prefi

So we only want to support one `protocol_prefix` per bridge (eg: `IBC_osmosis`)? That seems limiting if a certain chain has multiple protocols for generating assetIDs (like Picasso staking pallet generating share tokens). [This section](https://github.com/paritytech/xcm-format#concrete-identifiers) provides possible ways the IDs may be generated.

> Protocol is know to both parties as it was created during handshare (connection open, contract deploy, governance voting, whatever)

So we want to have a separate agreement instead of something like https://github.com/paritytech/xcm-format#6-universal-asset-identifiers which already solves this problem?

> So on top of that hashes do not need for predictability, .... Why it is better?

I think governance oriented process are more complicated as we see already in Polkadot governance. Specially when the alternative is to use universal asset identifiers.

Reply:
If stake/purchase is required, we'd be able to mitigate concerns about attacks on the AR storage

Reply:
> I think governance oriented process are more complicated as we see already in Polkadot governance. Specially when the alternative is to use automated universal asset identifiers.

Governance happens in any case. XCM channels and IBC doing client/connection are not permissionless.   So there is governance regardless of solution.

Reply:
With hashing foreign what is purpose/utility/value/usage of protocol prefix for locals?

Reply:
> This would just be one aspect I guess. There is also the aspect of preventing or authenticating that an asset received is actually the registered asset to prevent incorrect mapping or denial of service for registering assets. Say we received hypothetically 200 of IBC Osmosis chain asset of assetId=USDT. Then Picasso should understand how to correctly map it to already registered asset in the registry, say IBC/Osmosis/USDT. The channel which the asset arrives on should provide enough detail to map instead of being arbitrary. This is like signature verification of incoming messages on a channel to verify if a message is coming from the said source.

Why hashing is required here? I see that there is no need for implementing possibly desired behavior above.

Reply:
Imho if there is split in foreign vs local asset space, or attempt to put protocol and hash assets into single storage, i do not see why hashing needed. Its overengineered. Please keep it simple.

Reply:
As mentioned before, IBC-ICS20 and XCM requirements to map assets to/from remote to/from local. Mapping was shown on one of meetings, PRed in GH and chatted in Slack.  There is no need for hash.

Also XCVM was not fully discussed, as there is not clarity on this, I found, reported into Slack, that XCVM can use protocol id+denom if it wants local id to be well know by asset issuing origin.   Imho either XCVM should use own protocol (Local, Foreign, and Xcvm asset) or it can be protocol+denom. As XCVM is not aware neither of IBC-ICS20 nor XCM as it is transport agnostics. So it may use existing XCM/IBC-ICS20 assets, but as per what I know it will not use these protocols to issue assets.

Reply:
>Governance happens in any case. XCM channels and IBC doing client/connection are not permissionless. So there is governance regardless of solution.

Not for each new assetId used in the channel (we want this to be permissionless as per product requirements).

Reply:
> With hashing foreign what is purpose/utility/value/usage of protocol prefix for locals?

Locals still need not be coordinated by a single counter (ranges), namespacing by protocol_id(or anything else) allows each protocol to define their assetIds freely without central coordination by a pallet (eg: currency-factory).

Reply:
> Why hashing is required here? I see that there is no need for implementing possibly desired behavior above.

Hashing is to ensure both predictability (non-random) while ensuring uniqueness given a channel/protocol of a assetId.

Reply:
> Please keep it simple

Yes we want to, but need a usable solution for both cases not ignoring requirements. We want to achieve the simplest possible design that satisfies the requirements.

Reply:
> Mapping was shown on one of meetings, PRed in GH and chatted in Slack.

Any RFC/Design/Link/Reference? Why is this hidden information? ;)

> There is no need for hash, determinism is built in into protocols regardless of local assets representation.

Hash provides a way to avoid unintended/intended collisions when generating IDs without central coordination. Was reading about IBC [denoms](https://interchainacademy.cosmos.network/tutorials/5-ibc-dev/) (thanks @blasrodri) . There it also uses hashing for keeping the size limited. My idea is that for now mapping/storing the [hashed id ->protocol representation(xcm mulltiasset/IBC denom) of the foreign asset is possibly the best way to manage foreign asset registrations while avoiding collisions/security issues. 

> So it may use existing XCM/IBC-ICS20 assets, but as per what I know it will not use these protocols to issue assets.

I think we have some freedom for this future decisions as long as we have a general enough system that handles foreign and local assets in a well known/consistent/secure way.

Reply:
Nothing was hidden, people were invited, PR to GH was done. Slack chat was with Picasso teams representatives and Karel was at least notified. So not sure what you mean about hidden?

Reply:
> There it also uses hashing for keeping the size limited.
They store full prefixes and denoms afaik. Other they will not be able ICS20 transfers.

Reply:
> well known/consistent/secure way.

What you mean by this?

Reply:
> Not for each new assetId used in the channel (we want this to be permissionless as per product requirements in the coming releases).

What are hidden requirements? Just joking.

Ok. Sure. Things are fine - will work. 

Are products aware of how XCM/IBC-ICS20 works? :)

Reply:
> Locals still need not be coordinated by a single counter (ranges), namespacing by protocol_id(or anything else) allows each protocol to define their assetIds freely without central coordination by a pallet (eg: currency-factory).

What do you mean by central coordination? Where are hidden requirements for this?   What purpose that serves for?

Reply:
> Hashing is to ensure both predictability (non-random) while ensuring uniqueness given a channel/protocol of a assetId.

Sure. Still not clear why hashing is required? What for?

All works fine and better without.

Reply:
> Yes we want to, but need a usable solution for both cases not ignoring requirements. We want to achieve the simplest possible design that satisfies the requirements.

What """requirments""" are you referring? I have no idea. Seems like some hidden knowledge.

Reply:
Requirements are being discussed transparently in initiatives as per engineering handbook. You may refer IBC specific ones in the [initiatives](https://app.clickup.com/20465559/v/l/4-66217053-1) space in clickup. I recommend reading and commenting through them whenever you get a moment. The one that directly affects asset registration is INIT-59

Reply:
> Sure. Still not clear why hashing is required? What for? 

I'm tired of explaining the same thing when you don't seem to read what I explained above :) 

Anyway for the last time

"Hashing is to ensure both predictability (non-random) while ensuring uniqueness given a channel/protocol of a assetId."

Reply:
>What do you mean by central coordination? Where are hidden requirements for this? What purpose that serves for?

Central coordination when there is a one process/storage/code that is responsible to hand out the next asset ID in the system. Non-central coordination example can be found in ETH where each smart contract address generated can act as an assetID.

Reply:
> I'm tired of explaining the same thing when you don't seem to read what I explained above :)
Same thing on my side.

Reply:
> "Hashing is to ensure both predictability (non-random) while ensuring uniqueness given a channel/protocol of a assetId.

Hashing is not needed for predictability and uniqueness.....

Reply:
> Central coordination is when there is a one process/storage/code that is responsible to hand out the next asset ID in the system. 

In ETH one storage and one process responsible to handout id. Or it would not work.

>  Non-central coordination example can be found in ETH where each smart contract address generated can act as an assetID.

So XCM and IBC-ICS20 solve this without central coordination and without hashing. Nor should we. XCVM on other hand afaik was designed for central coordination.

Reply:
> I'm tired of explaining the same thing when you don't seem to read what I explained above :)

Why such assumption? I have read. I found you say nonsense.

Reply:
What I did not found RFC md file describes exact process of producing asset ids.

Reply:
btw ETH has 20 bytes address, not 16. and other chains has 32 bytes. we would have CW which uses 32 addresses. not 16.